apis/r/R/SOMAExperimentAxisQuery.R

```diff
@@ -994,6 +1132,233 @@ SOMAExperimentAxisQuery <- R6::R6Class(
+          soma_reader <- self$obs
+          soma_joinids <- self$obs_joinids()
+        },
+        var = {
+          soma_df <- self$var_df
+          soma_reader <- self$var
+          soma_joinids <- self$var_joinids()
+        }
+      )
+      ids <- if (is.null(index)) {
+        paste0(df_name, soma_joinids$as_vector())
+      } else {
+        index <- match.arg(arg = index, choices = soma_df$attrnames())
+        soma_reader(index)$concat()$GetColumnByName(index)$as_vector()
+      }
+      if (isTRUE(column_names)) {
+        column_names <- NULL
+      }
+      column_names <- column_names %||% setdiff(x = soma_df$attrnames(), y = index)
+      obj <- if (isFALSE(column_names) || rlang::is_na(column_names)) {
+        as.data.frame(matrix(nrow = length(ids), ncol = 0L))
+      } else {
+        as.data.frame(soma_reader(column_names)$concat()$to_data_frame())
+      }
+      row.names(obj) <- ids
+      return(obj)
+    },
+    .load_m_axis = function(layer, m_axis = c('obsm', 'varm'), type = "Embeddings") {
+      stopifnot(
+        is_scalar_character(layer),
+        is.character(m_axis),
+        is_scalar_character(type)
+      )
+      m_axis <- match.arg(arg = m_axis)
+      switch(
+        EXPR = m_axis,
+        obsm = {
+          soma_collection <- self$ms$obsm
+          soma_joinids <- self$obs_joinids()
+        },
+        varm = {
+          soma_collection <- self$ms$varm
+          soma_joinids <- self$var_joinids()
+        }
+      )
+      soma_axis <- soma_collection$get(layer)
+      mat <- if (inherits(soma_axis, 'SOMASparseNDArray')) {
+        as.matrix(self$to_sparse_matrix(collection = m_axis, layer_name = layer))
+      } else if (inherits(soma_axis, 'SOMADenseNDArray')) {
+        warning(
+          paste(
+            strwrap(paste(
+              type,
+              "for",
+              sQuote(layer),
+              "are encoded as dense instead of sparse; all arrays should be saved as",
+              sQuote('SOMASparseNDArrays')
+            )),
+            collapse = '\n'
+          ),
+          call. = FALSE,
+          immediate. = TRUE
+        )
+        ncoords <- max(tiledb::tiledb_array_get_non_empty_domain_from_name(
+          arr = soma_axis$object,
+          name = soma_axis$dimnames()[2L]
+        ))
+        coords <- list(
+          soma_joinids$as_vector(),
+          seq.int(0L, as.integer(ncoords))
+        )
+        soma_axis$read_dense_matrix(coords)
+      } else {
+        stop("Unknown SOMA Array type: ", class(soma_axis)[1L], call. = FALSE)
+      }
+      return(mat)
+    },
+    .load_p_axis = function(layer, p_axis = c('obsp', 'varp'), repr = c('C', 'T', 'R', 'D')) {
+      stopifnot(
+        is_scalar_character(layer),
+        is.character(p_axis),
+        is.character(repr)
+      )
+      p_axis <- match.arg(p_axis)
+      repr <- match.arg(repr)
+      mat <- self$to_sparse_matrix(collection = p_axis, layer_name = layer)
+      return(switch(
```

Comment:
Let's add a log statement prior to converting [CR]sparseMatrix to make it clear where the time is being spent.